miraclelinux/meta-emlinux
# cve-check: Get kernel patched/unpatched CVE information from cip-kernel-sec

# Purpose of pull request

Recent kernel CVE entries don't contain CPE in the NVD database. Therefore, cve-check cannot get these CVE information from NVD database. To report these CVEs, get CVE information by cip-kernel-sec that report all CVE status. Previous change only report unpatched CVEs only. This change reports patched CVEs as well.

# Test
## How to test

1. Run cve-check for linux-base and linux-k510 using warrior branch
2. Run cve-check for linux-base and linux-k510 with this PR's commit
3. Compare above results that only number of PATCHED CVEs is updated

Here is check script.

```python
#!/usr/bin/env python3

import sys
import os.path
import pprint

def line_value(line):
    return line.split(":")[1].strip()

def parse_cvefile(cvefile):
    cves = {}

    with open(cvefile) as f:
        lines = f.readlines()
        cve = None
        cvssv3 = None
        status = None

        i = 0
        while True:
            if i >= len(lines):
                break
            line = lines[i]

            if line.startswith("CVE:"):
                cve = line_value(line)
            elif line.startswith("CVSS v3 BASE SCORE:"):
                cvssv3 = line_value(line)
            elif line.startswith("CVE STATUS:"):
                status = line_value(line)
            elif line.startswith("MORE INFORMATION:"):
                if not cve is None:
                    if not cve in cves:
                        cves[cve] = {
                            "CVE": cve,
                            "CVSSv3": cvssv3,
                            "STATUS": status,
                        }
                        cve = None
                        cvssv3 = None
                        status = None
                    else:
                        print(f"CVE {cve} is duplicated")
                        exit(1)

            i += 1

    return cves

def cve_summary(cves):
    patched = 0
    unpatched = 0
    unkown = 0

    for k in cves:
        status = cves[k]["STATUS"]
        if status == "Patched":
            patched += 1
        elif status == "Unpatched":
            unpatched += 1
        else:
            unkown += 1

    return patched, unpatched, unkown

def get_unpatched_cves(cves):
    ret = []

    for k in cves:
        if cves[k]["STATUS"] == "Unpatched":
            ret.append(k)

    return sorted(ret)

def show_unpatched_cves(old_cves, new_cves):
    old_unpatched = get_unpatched_cves(old_cves)
    new_unpatched = get_unpatched_cves(new_cves)

    if len(old_unpatched) > len(new_unpatched):
        diff = sorted(list(set(old_unpatched) - set(new_unpatched)))
    else:
        diff = sorted(list(set(new_unpatched) - set(old_unpatched)))

    if len(diff) > 0:
        print("CVEs that only in new data")
        for cve in diff:
            print(cve)

def main(cvefile_old, cvefile_new):
    old_cves = parse_cvefile(cvefile_old)
    new_cves = parse_cvefile(cvefile_new)

    old_patched, old_unpatched, old_unknown = cve_summary(old_cves)
    new_patched, new_unpatched, new_unknown = cve_summary(new_cves)

    print(f"{cvefile_old} cve-check result:")
    print(f"  Total CVEs: {len(old_cves)}")
    print(f"  patched: {old_patched}")
    print(f"  unpatched: {old_unpatched}")
    print(f"  unknown status: {old_unknown}")
    print(f"{cvefile_new} cve-check result:")
    print(f"  Total CVEs: {len(new_cves)}")
    print(f"  patched: {new_patched}")
    print(f"  unpatched: {new_unpatched}")
    print(f"  unknown status: {new_unknown}")

    show_unpatched_cves(old_cves, new_cves)

if __name__ == "__main__":
    if not len(sys.argv) == 3:
        print(f"Usage: {sys.argv[0]} <file> <file>")
        exit(1)
        
    main(sys.argv[1], sys.argv[2])
```

## Test result

As you can see only number of patched cve count is increased.

```
$ ./parse_cve_log.py linux-base.warrior linux-base.new 
linux-base.warrior cve-check result:
  Total CVEs: 4225
  patched: 3574
  unpatched: 651
  unknown status: 0
linux-base.new cve-check result:
  Total CVEs: 5998
  patched: 5347
  unpatched: 651
  unknown status: 0
$ ./parse_cve_log.py linux-k510.warrior linux-k510.new 
linux-k510.warrior cve-check result:
  Total CVEs: 4150
  patched: 3623
  unpatched: 527
  unknown status: 0
linux-k510.new cve-check result:
  Total CVEs: 6004
  patched: 5477
  unpatched: 527
  unknown status: 0

```
